File: src/PicLens.Core/Domain/ListItemSorter.cs
using PicLens.Core.Models;
using CoreSortKey = PicLens.Core.Models.SortKey;

namespace PicLens.Core.Domain;

public static class ListItemSorter
{
    public static IReadOnlyList<ListItem> Sort(
        IEnumerable<ListItem> items,
        SortState sort,
        bool keepFoldersFirst)
    {
        var sorted = items.ToList();
        sorted.Sort((left, right) =>
        {
            if (left is null)
            {
                return right is null ? 0 : -1;
            }

            if (right is null)
            {
                return 1;
            }

            if (keepFoldersFirst && left.GetType() != right.GetType())
            {
                return left is FolderListItem ? -1 : 1;
            }

            var result = sort.Key == CoreSortKey.Name
                ? NaturalCompare(left.Name, right.Name)
                : (left.ModifiedAtMs ?? 0).CompareTo(right.ModifiedAtMs ?? 0);

            return sort.Direction == SortDirection.Asc ? result : -result;
        });

        return sorted;
    }

    private static int NaturalCompare(string left, string right)
    {
        var leftIndex = 0;
        var rightIndex = 0;

        while (leftIndex < left.Length && rightIndex < right.Length)
        {
            var leftChar = left[leftIndex];
            var rightChar = right[rightIndex];

            if (char.IsDigit(leftChar) && char.IsDigit(rightChar))
            {
                var result = CompareNumberRuns(left, ref leftIndex, right, ref rightIndex);
                if (result != 0)
                {
                    return result;
                }

                continue;
            }

            var charResult = char.ToUpperInvariant(leftChar).CompareTo(char.ToUpperInvariant(rightChar));
            if (charResult != 0)
            {
                return charResult;
            }

            charResult = leftChar.CompareTo(rightChar);
            if (charResult != 0)
            {
                return charResult;
            }

            leftIndex++;
            rightIndex++;
        }

        return (left.Length - leftIndex).CompareTo(right.Length - rightIndex);
    }

    private static int CompareNumberRuns(string left, ref int leftIndex, string right, ref int rightIndex)
    {
        var leftStart = leftIndex;
        while (leftIndex < left.Length && char.IsDigit(left[leftIndex]))
        {
            leftIndex++;
        }

        var rightStart = rightIndex;
        while (rightIndex < right.Length && char.IsDigit(right[rightIndex]))
        {
            rightIndex++;
        }

        var leftSignificant = FirstSignificantDigit(left, leftStart, leftIndex);
        var rightSignificant = FirstSignificantDigit(right, rightStart, rightIndex);
        var leftSignificantLength = leftIndex - leftSignificant;
        var rightSignificantLength = rightIndex - rightSignificant;

        var lengthResult = leftSignificantLength.CompareTo(rightSignificantLength);
        if (lengthResult != 0)
        {
            return lengthResult;
        }

        for (var i = 0; i < leftSignificantLength; i++)
        {
            var digitResult = left[leftSignificant + i].CompareTo(right[rightSignificant + i]);
            if (digitResult != 0)
            {
                return digitResult;
            }
        }

        var runLengthResult = (rightIndex - rightStart).CompareTo(leftIndex - leftStart);
        if (runLengthResult != 0)
        {
            return runLengthResult;
        }

        return left.AsSpan(leftStart, leftIndex - leftStart)
            .SequenceCompareTo(right.AsSpan(rightStart, rightIndex - rightStart));
    }

    private static int FirstSignificantDigit(string value, int start, int end)
    {
        var index = start;
        while (index < end - 1 && value[index] == '0')
        {
            index++;
        }

        return index;
    }
}

File: qt/src/core/src/list_item_sorter.cpp
#include <piclens/core/list_item_sorter.h>

#include <QStringView>

#include <algorithm>

namespace piclens::core::list_item_sorter {
namespace {

int firstSignificantDigit(const QString &value, int start, int end)
{
    int index = start;
    while (index < end - 1 && value.at(index) == QLatin1Char('0')) {
        ++index;
    }
    return index;
}

int compareNumberRuns(
    const QString &left,
    int &leftIndex,
    const QString &right,
    int &rightIndex)
{
    const int leftStart = leftIndex;
    while (leftIndex < left.size() && left.at(leftIndex).isDigit()) {
        ++leftIndex;
    }

    const int rightStart = rightIndex;
    while (rightIndex < right.size() && right.at(rightIndex).isDigit()) {
        ++rightIndex;
    }

    const int leftSignificant = firstSignificantDigit(left, leftStart, leftIndex);
    const int rightSignificant = firstSignificantDigit(right, rightStart, rightIndex);
    const int leftSignificantLength = leftIndex - leftSignificant;
    const int rightSignificantLength = rightIndex - rightSignificant;

    if (leftSignificantLength != rightSignificantLength) {
        return leftSignificantLength < rightSignificantLength ? -1 : 1;
    }

    for (int index = 0; index < leftSignificantLength; ++index) {
        const ushort leftDigit = left.at(leftSignificant + index).unicode();
        const ushort rightDigit = right.at(rightSignificant + index).unicode();
        if (leftDigit != rightDigit) {
            return leftDigit < rightDigit ? -1 : 1;
        }
    }

    const int leftRunLength = leftIndex - leftStart;
    const int rightRunLength = rightIndex - rightStart;
    if (leftRunLength != rightRunLength) {
        return rightRunLength < leftRunLength ? -1 : 1;
    }

    return QStringView(left).sliced(leftStart, leftRunLength)
        .compare(QStringView(right).sliced(rightStart, rightRunLength), Qt::CaseSensitive);
}

int naturalCompare(const QString &left, const QString &right)
{
    int leftIndex = 0;
    int rightIndex = 0;

    while (leftIndex < left.size() && rightIndex < right.size()) {
        const QChar leftCharacter = left.at(leftIndex);
        const QChar rightCharacter = right.at(rightIndex);

        if (leftCharacter.isDigit() && rightCharacter.isDigit()) {
            const int result = compareNumberRuns(left, leftIndex, right, rightIndex);
            if (result != 0) {
                return result;
            }
            continue;
        }

        const ushort leftUpper = leftCharacter.toUpper().unicode();
        const ushort rightUpper = rightCharacter.toUpper().unicode();
        if (leftUpper != rightUpper) {
            return leftUpper < rightUpper ? -1 : 1;
        }

        if (leftCharacter != rightCharacter) {
            return leftCharacter.unicode() < rightCharacter.unicode() ? -1 : 1;
        }

        ++leftIndex;
        ++rightIndex;
    }

    const int leftRemaining = left.size() - leftIndex;
    const int rightRemaining = right.size() - rightIndex;
    if (leftRemaining == rightRemaining) {
        return 0;
    }
    return leftRemaining < rightRemaining ? -1 : 1;
}

} // namespace

QVector<ListItem> sort(
    const QVector<ListItem> &items,
    SortState sortState,
    bool keepFoldersFirst)
{
    QVector<ListItem> sorted = items;
    std::sort(sorted.begin(), sorted.end(), [&](const ListItem &left, const ListItem &right) {
        if (keepFoldersFirst && isFolder(left) != isFolder(right)) {
            return isFolder(left);
        }

        int result = 0;
        if (sortState.key == SortKey::Name) {
            result = naturalCompare(itemName(left), itemName(right));
        } else {
            const qint64 leftModified = itemModifiedAtMs(left).value_or(0);
            const qint64 rightModified = itemModifiedAtMs(right).value_or(0);
            result = leftModified < rightModified ? -1 : leftModified > rightModified ? 1 : 0;
        }

        return sortState.direction == SortDirection::Asc ? result < 0 : result > 0;
    });
    return sorted;
}

QString itemName(const ListItem &item)
{
    return std::visit([](const auto &value) { return value.name; }, item);
}

QString itemPath(const ListItem &item)
{
    return std::visit([](const auto &value) { return value.path; }, item);
}

std::optional<qint64> itemModifiedAtMs(const ListItem &item)
{
    return std::visit([](const auto &value) { return value.modifiedAtMs; }, item);
}

bool isFolder(const ListItem &item)
{
    return std::holds_alternative<FolderListItem>(item);
}

} // namespace piclens::core::list_item_sorter
